### PR TITLE
Regexp hooked vars

### PR DESCRIPTION
**Pull request by aardvark179**
This moves the regexp special variables to Ruby. I haven't converted all uses of the `regexp_set_last_match` primitive to the `RegexpOperations.set_last_match` method as I'm still exploring ways to default the binding argument.

### CODE REVIEW

File: src/main/ruby/core/truffle/regexp_operations.rb [hunk 1]
```diff
@@ -19,5 +19,19 @@ def self.match(re, str, pos=0)
...
+    def self.last_match(a_binding)
+      Truffle::KernelOperations.frame_local_variable_get(:$~, a_binding)
+    end
+
+    def self.set_last_match(value, a_binding)
+      unless value.nil? || Truffle::Type.object_kind_of?(value, MatchData)
+        raise TypeError, "Wrong argument type #{value} (expected MatchData)"
+      end
+      Truffle::KernelOperations.frame_local_variable_set(:$~, a_binding, value)
+    end
```
Comment (eregon) [inline src/main/ruby/core/truffle/regexp_operations.rb]: Is this code used somewhere else?
Otherwise I think it would make sense to have it in the lambdas above directly.

File: src/main/ruby/core/regexp.rb [hunk 2]
```diff
@@ -396,3 +396,8 @@ def to_s
     self[0]
   end
 end
+
+Truffle::KernelOperations.define_hooked_variable(
+  :$~,
+  Truffle::Graal.always_split( -> b { Truffle::RegexpOperations.last_match(b) }),
+  Truffle::Graal.always_split( -> v, b { Truffle::RegexpOperations.set_last_match(v, b) }))
```
Comment (eregon) [inline src/main/ruby/core/regexp.rb]: Should not be a space after `(`.

File: src/main/ruby/core/regexp.rb [hunk 3]
```diff
@@ -401,3 +401,19 @@ def to_s
   :$~,
   Truffle::Graal.always_split( -> b { Truffle::RegexpOperations.last_match(b) }),
   Truffle::Graal.always_split( -> v, b { Truffle::RegexpOperations.set_last_match(v, b) }))
+
+Truffle::KernelOperations.define_hooked_variable_with_is_defined(
+  :$`,
```
Comment (eregon) [inline src/main/ruby/core/regexp.rb]: This one is good at killing syntax highlighting, maybe try :"$`"?

File: src/main/java/org/truffleruby/core/regexp/RegexpNodes.java [hunk 4]
```diff
@@ -345,76 +345,6 @@ public static boolean frameIsNotSend(RubyContext context, Object callerFrame) {
...
-
-            final DynamicObject matchData = matchNode.execute(regexp, string, matcher, 0, range, false);
-            setLastMatchNode.executeSetLastMatch(frame, matchData);
-
-            if (matchData != nil()) {
-                return matcher.getBegin();
-            }
-
-            return nil();
-        }
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/regexp/RegexpNodes.java]: This part of the behavior is already handled in `regexp_search_region` it seems.
It would be nice to move the comment there.
And I think that `dup` of the source String should actually be moved to `MatchNode` eventually since that node creates the MatchData.

File: src/main/java/org/truffleruby/parser/BodyTranslator.java [hunk 5]
```diff
@@ -273,15 +274,13 @@
...
-import java.util.Deque;
-import java.util.Iterator;
-import java.util.List;
+import com.oracle.truffle.api.CallTarget;
+import com.oracle.truffle.api.Truffle;
+import com.oracle.truffle.api.frame.FrameSlot;
+import com.oracle.truffle.api.nodes.NodeUtil;
+import com.oracle.truffle.api.object.DynamicObject;
+import com.oracle.truffle.api.source.Source;
+import com.oracle.truffle.api.source.SourceSection;
```
Comment (eregon) [inline src/main/java/org/truffleruby/parser/BodyTranslator.java]: Changing the `import` order is annoying when rebasing, try to avoid as possible.

File: src/main/ruby/core/regexp.rb [hunk 6]
```diff
@@ -417,3 +417,19 @@ def to_s
...
+    -> b { match = Truffle::RegexpOperations.last_match(b)
+           match[0] if match }),
+  -> { raise SyntaxError, "Can't set variable $&"},
+  Truffle::Graal.always_split( -> b { 'global-variable' if Truffle::RegexpOperations.last_match(b) }))
+
+Truffle::KernelOperations.define_hooked_variable_with_is_defined(
+  :$+,
+  Truffle::Graal.always_split(
+    -> b { match = Truffle::RegexpOperations.last_match(b)
+           match.captures.reject { |m| m.nil? }.last if match }),
```
Comment (eregon) [inline src/main/ruby/core/regexp.rb]: `Enumerable#find`, no?

File: src/main/java/org/truffleruby/core/proc/ProcNodes.java [hunk 7]
```diff
@@ -205,7 +205,7 @@ public int arity(DynamicObject proc) {
         @Specialization
         public DynamicObject binding(DynamicObject proc) {
             final MaterializedFrame frame = Layouts.PROC.getDeclarationFrame(proc);
-            return BindingNodes.createBinding(getContext(), frame);
+            return frame == null ? nil() : BindingNodes.createBinding(getContext(), frame);
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/proc/ProcNodes.java]: This might have bad performance effects, like preventing EA of the Binding.
Is it needed?

File: src/main/ruby/core/regexp.rb [hunk 8]
```diff
@@ -138,6 +138,16 @@ def self.compatible?(*patterns)
     end
   end
 
+  def self.last_match(index=nil)
+    match = Truffle::RegexpOperations.last_match(Truffle.invoke_primitive(:caller_binding))
+    if index
+      Truffle::Type.coerce_to index, Fixnum, :to_int
```
Comment (eregon) [inline src/main/ruby/core/regexp.rb]: index = ...

File: src/main/ruby/core/regexp.rb [hunk 9]
```diff
@@ -381,3 +406,40 @@ def to_s
...
+  -> { raise SyntaxError, "Can't set variable $&"},
+  Truffle::Graal.always_split( -> b { 'global-variable' if Truffle::RegexpOperations.last_match(b) }))
+
+Truffle::KernelOperations.define_hooked_variable_with_is_defined(
+  :$+,
+  Truffle::Graal.always_split(
+    -> b { match = Truffle::RegexpOperations.last_match(b)
+           match.captures.reject { |m| m.nil? }.last if match }),
+  -> { raise SyntaxError, "Can't set variable $+"},
+  Truffle::Graal.always_split( -> b { 'global-variable' if Truffle::RegexpOperations.last_match(b) }))
```
Comment (eregon) [inline src/main/ruby/core/regexp.rb]: The splitting is very verbose.
Maybe always split in `define_hooked_variable_with_is_defined`?
Or at least when it takes a binding argument (arity 1)?
Otherwise you could also define a helper in a local var like
```ruby
split = -> proc { Truffle::Graal.always_split(proc) }
split.(-> b { ... })
```

**Comment by eregon** [comment]
> I haven't converted all uses of the `regexp_set_last_match` primitive to the `RegexpOperations.set_last_match` method as I'm still exploring ways to default the binding argument.

I think passing the Binding is a very natural way to do it and involves no magic.

File: src/main/java/org/truffleruby/core/proc/ProcNodes.java (hunk 7 above)
Comment (aardvark179) [reply src/main/java/org/truffleruby/core/proc/ProcNodes.java]: Blocks created from methods don't have a declaration frame. I could change it to two specializations and a guard if that would be better.

Comment (eregon) [reply src/main/java/org/truffleruby/core/proc/ProcNodes.java]: Do you have such a case a Method#to_proc in this PR?
It might be easier/cleaner to just give a declaration frame for Method#to_proc for consistency.
Two specializations would definitely be better.

File: src/main/ruby/core/truffle/regexp_operations.rb (hunk 1 above)
Comment (aardvark179) [reply src/main/ruby/core/truffle/regexp_operations.rb]: It is/will be used elsewhere, because we can now get rid of the `regexp_set_block_last_match` and `regexp_set_last_match` primitives which are used in various string methods..

File: src/main/java/org/truffleruby/core/regexp/RegexpNodes.java (hunk 4 above)
Comment (aardvark179) [reply src/main/java/org/truffleruby/core/regexp/RegexpNodes.java]: I agree. I think it wasn't in match node originally because it was the wrong side of a truffle boundary. I'll move the comment for now

File: src/main/java/org/truffleruby/core/proc/ProcNodes.java (hunk 7 above)
Comment (aardvark179) [reply src/main/java/org/truffleruby/core/proc/ProcNodes.java]: Just checked using `irb` and in MRI procs always have a binding, so I'd better fix that so we match.

File: src/main/ruby/core/truffle/regexp_operations.rb (hunk 1 above)
Comment (eregon) [reply src/main/ruby/core/truffle/regexp_operations.rb]: OK.

File: src/main/java/org/truffleruby/core/proc/ProcNodes.java (hunk 7 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/proc/ProcNodes.java]: So then the solution is to create a declaration frame in Method#to_proc too then?
If this is not needed in this PR it would be nicer to extract in its own PR.

File: src/main/ruby/core/regexp.rb [hunk 10]
```diff
@@ -192,6 +202,21 @@ def initialize(pattern, opts=nil, lang=nil)
...
+      return nil
+    end
+    result = Truffle::RegexpOperations.match(self, str, 0)
+    Truffle.invoke_primitive(:regexp_set_last_match, result)
+
+    if result
+      result.begin 0
+    else
+      nil
+    end
```
Comment (eregon) [inline src/main/ruby/core/regexp.rb]: This can be one line:
```ruby
result.being(0) if result
```
But if you find this clearer it's fine too.

File: src/main/ruby/core/regexp.rb [hunk 11]
```diff
@@ -381,3 +406,36 @@ def to_s
...
   end
 end
+
+Truffle::KernelOperations.define_hooked_variable(
+  :$~,
+  -> b { Truffle::RegexpOperations.last_match(b) },
+  -> v, b { Truffle::RegexpOperations.set_last_match(v, b) })
+
+Truffle::KernelOperations.define_hooked_variable(
+  :$`,
```
Comment (eregon) [inline src/main/ruby/core/regexp.rb]: Could you try quoting this and the `$'`? That would likely fix the bad highlighting at least on GitHub.

File: src/main/java/org/truffleruby/parser/TranslatorEnvironment.java [hunk 12]
```diff
@@ -161,11 +160,7 @@ public ReadLocalNode findFrameLocalGlobalVarNode(String name, Source source, Sou
         final FrameSlot slot = methodScope.declareVar(name);
 
         final LocalVariableType type;
-        if (GlobalVariables.ALWAYS_DEFINED_GLOBALS.contains(name)) {
-            type = LocalVariableType.ALWAYS_DEFINED_GLOBAL;
-        } else {
-            type = LocalVariableType.FRAME_LOCAL_GLOBAL;
-        }
+        type = LocalVariableType.FRAME_LOCAL_GLOBAL;
```
Comment (eregon) [inline src/main/java/org/truffleruby/parser/TranslatorEnvironment.java]: Can be one line with the declaration of the local var, or even inlined below.
`GlobalVariables.ALWAYS_DEFINED_GLOBALS` can also be removed I'd guess (can be another PR).

File: src/main/ruby/core/truffle/regexp_operations.rb [hunk 13]
```diff
@@ -19,5 +19,19 @@ def self.match(re, str, pos=0)
...
+
+    Truffle::Graal.always_split(method(:last_match))
+    Truffle::Graal.always_split(method(:set_last_match))
```
Comment (eregon) [inline src/main/ruby/core/truffle/regexp_operations.rb]: Please keep method modifiers (private, always_split, etc) just below the method definition.
Otherwise it's hard to see last_match is always split.

**Comment by pitr-ch** [comment]
@aardvark179  looks good 👍
